### ISSUE TITLE
Problem with grad_overrides in OpFromGraph

### ISSUE THREAD

**Issue by mtomassoli** [Theano/Theano#5685]
There seems to be a problem with OpFromGraph when the user's gradient function uses a variable which doesn't belong to the main graph. Here's a minimal sample to recreate the bug:

```
import theano
import theano.tensor as T
from theano.sandbox.rng_mrg import MRG_RandomStreams as RandomStreams
import numpy as np

floatX = theano.config.floatX


##############################
it_works = False
##############################

srng = RandomStreams(np.random.randint(1, 2147462579))


def my_grad(inputs, out_grads):
    X, = inputs
    R = srng.uniform(X.shape, dtype=floatX)
    return [X + R]

X = T.vector('X', dtype=floatX)

res = X.mean(axis=0)

if it_works:
    op = theano.OpFromGraph([X], [res])
    op.grad = my_grad
else:
    op = theano.OpFromGraph([X], [res], inline=True, grad_overrides=my_grad)

f = theano.function([X], T.grad(op(X), [X]))

X_np = np.ones(1000, dtype=floatX)
f(X_np)

```

**Comment by nouiz** [comment]
This isn't supported. The grad should only use the variables from inputs
and out_grads in this example.

I know people have done something similar, to include random numbers during
the gradient. But I forgot how they did it.

@lamblin, do you remember how this can be done?

**Comment by khaotik** [comment]
It's possible to define standalone function for gradient:

```
X = T.vector('x')
R = srng.uniform(size=T.shape(s_x), low=-1., high=1., dtype=floatX)
f_grad = th.function([X], X + R)
```

So, in theory, the example code *should* work. I'll look into this.

For a quick workaround, you can give a fixed `size` to `srng.random`, if it's know in advance. Or you can apply noise to standard `grad` result, not using gradient override at all.

**Comment by khaotik** [comment]
The problem isn't gradient override. OpFromGraph + rng -> bug. A even smaller example:

```
import theano as th
import theano.tensor as T
from theano.sandbox.rng_mrg import MRG_RandomStreams as RandomStreams
import numpy as np

floatX = th.config.floatX

srng = RandomStreams(np.random.randint(1, 2147462579))

s_x = T.vector('x')
R = srng.uniform(size=T.shape(s_x), low=-1., high=1., ndim=1, dtype=floatX)
op_grad = th.OpFromGraph([s_x], [s_x + R])

x = T.vector()
fn = th.function([x], op_grad(x))

X_np = np.ones(1000, dtype=floatX)
fn(X_np)
```

**Comment by lamblin** [comment]
Indeed, OpFromGraph does not handle default_updates correctly (yet).
Something more to add to the OpFromGraph wishlist...

**Comment by Sentient07** [comment]
Any fix/workaround for this yet? I am having the same problem, with using rng with OpGraph's custom gradient.

**Comment by nouiz** [comment]
Can you confirm you use random numbers? I don't recall random numbers in the gumbel softmax.

**Comment by Sentient07** [comment]
I realized later today that it was due to the temperature being a shared variable in my code. I just changed it to a `fscalar` and things are fine now. Sorry about this

**Comment by Sentient07** [comment]
I didn't actually check with the number of variables that are being inputed to the node. There seems to be one extra variable that is taken as input. The desired is 3, but the grad method seems to take 4. I am not sure why this is happening. I have attached a snippet of my code and what pdb gives
<img width="1158" alt="screenshot at jun 16 18-08-25" src="https://user-images.githubusercontent.com/6195312/27226951-f7994192-52be-11e7-8502-c3bab397f38d.png">

**Comment by khaotik** [comment]
@Sentient07 I think the extra input is random state. At the 10th line, insert `print(theano.gof.graph.inputs(st_gsm))` to see them.

**Comment by Sentient07** [comment]
@khaotik But I think in that case, I would be required to differentiate the cost with respect to that random variable also, right? I don't think it'd be a correct estimate of the gradient

**Comment by lamblin** [comment]
Should we close this issue?

**Comment by Sentient07** [comment]
> Should we close this issue?

I don't think the issue is resolved yet. I wasn't able to overcome this problem. We need to add this to OpFromGraph wishlist as you said, if time permits.
